src/Visualization/Python/ApplyPointwise.py

```diff
@@ -62,11 +76,116 @@ def get_tensor_component_names(tensor_name: str,
     ]
 
 
+class KernelArg:
+    """Subclasses provide data for a kernel function argument"""
+    def get(self, element: Element, all_tensor_data: Dict[str, Tensor]):
+        """Get the data that will be passed to the kernel function"""
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class TensorArg(KernelArg):
+    """Kernel argument loaded from a volume file tensor dataset"""
+    tensor_type: Type[Tensor]
+    dataset_name: str
+
+    @cached_property
+    def component_names(self):
+        return get_tensor_component_names(self.dataset_name, self.tensor_type)
+
+    def get(self, element: Element, all_tensor_data: Dict[str, Tensor]):
+        return self.tensor_type(
+            np.asarray(all_tensor_data[self.dataset_name])[:,
+                                                           element.data_slice])
+
+
+@dataclass(frozen=True)
+class ElementArg(KernelArg):
+    """Kernel argument retrieved from an element in the computational domain"""
+    element_attr: str
+
+    def get(self, element: Element, all_tensor_data: Dict[str, Tensor]):
+        return getattr(element, self.element_attr)
+
+
+def parse_kernel_arg(arg: inspect.Parameter,
+                     map_input_names: Dict[str, str]) -> KernelArg:
+    """Determine how data for a kernel function argument will be retrieved
+
+    Examines the name and type annotation of the argument. The following
+    arguments are supported:
+
+    - Mesh: Annotate the argument with a 'Mesh' type.
+    - Coordinates: Annotate the argument with a 'tnsr.I' type and name it
+        "logical_coords" / "logical_coordinates" / "xi" or "inertial_coords" /
+        "inertial_coordinates" / "x".
+    - Jacobians: Annotate the argument with a 'Jacobian' or 'InverseJacobian'.
+    - Any tensor dataset: Annotate the argument with the tensor type. By
+        default, the argument name is transformed to CamelCase to determine the
+        dataset name in the volume data file. Specify a mapping in
+        'map_input_names' to override the default.
+
```

Comment:
Can you give an example of how one would for example compute the 1-index constraint in scalar wave? There's a lot to unpack here, and I think having an example would make that much easier to unpack.

How does this interact with C++ functions wrapped with pybind? Can you document that too? I'm not clear how I could call a function that takes `psi, mesh, inv_jac` and compute the 1-index constraint locally if I have a C++ bound function.

Reply:
Added some docs. To call a binding just pass something like `-k spectre.PointwiseFunctions.ScalarWave.one_index_constraint` to the CLI. Not sure how well that works for bindings that have multiple overloads, we'll see. Once a couple of suitable bindings are merged, like #4644 or #4637, I can add a test.

Reply:
Okay, sounds good!